### ISSUE TITLE
Roslyn workspace should share SyntaxTrees across project targets when possible.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#65278]
From talking with @jasonmalinowski we believe there would be a large memory savings here for solutions with projects that multi-target (e.g. netstandard2.0, netcoreapp3.1, etc.).  The cases that should be very safe to share are ones where the tree does not contain pp-directives, and where the lang-version for different targets are the same.

The reason for this is that pp-directives can parse differently depending on what PP directives are defined.  And lang-version can affect how certain constructs are parsed (like records), as well as affect what errors are reported (like if certain constructs are or aren't supported).

1. It would be valuable for us to collect some telemetry to see what percentage of projects *that are using multi-targetting* end up using the same lang-versions.  On top of that, it would be good to keep track of what percentage of their files do/do-not use pp-directives.
2. we currently do keep track of stuff like pp-directives for fast-switching of debug/release.  However, that tracking could be finer-grained.  I believe we check if there are *any* directives, where we really only care about `#if` directives.  e.g. `#region` and `#nullable` shouldn't change things here.
3. Roslyn would be a great test bed for this as we have many targets, and tons of files.  We could get a huge win here, that would translate out to projects similar to ours.
4. We may want to drive compiler APIs to assist us here.  For example, it would be great if we could cheaply just ask if a file contains pp-directives.  This could be a weak heuristic (potentially even at the source-text level).  For example, just asking if any line starts with `spaces-#-if`.  Or just allowing this to be determined from a green tree, etc.

**Comment by jasonmalinowski** [comment]
Thinking about this a bit:

> The cases that should be very safe to share are ones where the tree does not contain pp-directives, and where the lang-version for different targets are the same.

That language version constraint might be a challenge here: given https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version each of the different targets may have their own language version. Some projects (like Roslyn) just override the language version for it to be consistent, but that's not a supported behavior. @jaredpar and @CyrusNajmabadi I know in the past there were some places the compiler used the language version in parsing, but for any _recent_ language versions is that still the case? I recall somebody mentioning there was some desire to do that less and push the errors into semantic errors in the compiler for 'unsupported' stuff but not sure if that happened or not.

> I believe we check if there are any directives, where we really only care about #if directives. e.g. #region and #nullable shouldn't change things here.

A potential stronger optimization (although tricky, so we should only implement if it has clear value) would be to track at least at the high level which set of preprocessor symbols were involved in the #if that were seen. For multi-targeting, the only thing that might differ between the targets is the NETCOREAPP symbols, but a tree containing #if DEBUG still wouldn't be impacted either way.

**Comment by CyrusNajmabadi** [comment]
>  I know in the past there were some places the compiler used the language version in parsing, but for any recent language versions is that still the case?

Yes.  100%.  Records parse entirely different.  e.g. `record Foo() { }` is a method prior to records being supported, and it's a record with the right lang version.

It's also commonly used for parsing diagnostics (though i woudl support moving those all to the binder).

**Comment by CyrusNajmabadi** [comment]
> That language version constraint might be a challenge here: given https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version each of the different targets may have their own language version. 

ick :'(

**Comment by CyrusNajmabadi** [comment]
> A potential stronger optimization (although tricky, so we should only implement if it has clear value) would be to track at least at the high level which set of preprocessor symbols were involved in the #if that were seen. For multi-targeting, the only thing that might differ between the targets is the NETCOREAPP symbols, but a tree containing #if DEBUG still wouldn't be impacted either way.

Yup yup.  I had thought about that, but thought it likely too difficult.  But we should have all things on the table for consideration.  

PP-directives are a core part of the roslyn syntax model.  It would not be strange for there to be a really intelligent API exposed here to support this case.

**Comment by jaredpar** [comment]
My suspicion is that most code bases which significantly multi-target also explicitly set the language version. To do otherwise just ends up with a lot of hair pulling because you end up with code that compiles in one TF and fails in the other.

--
jaredpar
https://twitter.com/jaredpar
________________________________

**Comment by CyrusNajmabadi** [comment]
```c#
                case SyntaxKind.IdentifierToken:
                    if (CurrentToken.ContextualKind == SyntaxKind.RecordKeyword)
                    {
                        // This is an unusual use of LangVersion. Normally we only produce errors when the langversion
                        // does not support a feature, but in this case we are effectively making a language breaking
                        // change to consider "record" a type declaration in all ambiguous cases. To avoid breaking
                        // older code that is not using C# 9 we conditionally parse based on langversion
                        return IsFeatureEnabled(MessageID.IDS_FeatureRecords);
                    }
```

Behold its glory.

**Comment by 333fred** [comment]
> Yes. 100%. Records parse entirely different.

We also parse `required` a bit differently: we more strongly assume it's a modifier in C# 11, while in pre-11 we do a more pessimistic heuristic. And I honestly wouldn't expect these differences to get smaller. We _could_ consider setting a bit in the parser when a potential language version difference is encountered. I suspect that many, perhaps most, files wouldn't run into these types of differences, and those could be shared safely?

**Comment by CyrusNajmabadi** [comment]
@333fred that's a possible idea.  Though i have a feeling over time it will have diminishing returns.  For example, as people start to adopt newer features (records, required, and what comes next) it is more and more likely for files to contain at least one of these things.  

I'm hoping for what jared mentioned.  Specifically that in practice the lang versions are locked to the same thing.  Indeed, we could probably drive this with telemetry to find out if that's the case.  @jasonmalinowski do you thinhk existing telemetry might tell us this?  or would we need to add something specific?

**Comment by 333fred** [comment]
> Though i have a feeling over time it will have diminishing returns.

100% agreed.

**Comment by jasonmalinowski** [comment]
> @jasonmalinowski do you thinhk existing telemetry might tell us this? or would we need to add something specific?

I'm not aware of any telemetry off the top of my head for this, but the project system might have something.

>  We could consider setting a bit in the parser when a potential language version difference is encountered. I suspect that many, perhaps most, files wouldn't run into these types of differences, and those could be shared safely?

Rather than one bit you could imagine the parser could hand out two pieces of information: what would be the highest and lowest known language versions where the tree could still be reused. So to pseudocode the record example:

```
                case SyntaxKind.IdentifierToken:
                    if (CurrentToken.ContextualKind == SyntaxKind.RecordKeyword)
                    {
                        bool isRecordsEnabled = IsFeatureEnabled(MessageID.IDS_FeatureRecords);
                        if (isRecordsEnabled)
                        {
                            // We are parsing this as a C# record syntax, we cannot reuse this tree for anything lower than C# 9
                            _lowestCompatibleVersion = Math.Max(_lowestCompatibleVersion, LangVersion.9)
                            return true;
                        }
                       else
                       {
                            // Records are not enabled, so we're parsing this the old way.
                            _highestCompatibleVersion = Math.Min(_highestCompatibleVersion, LangVersion.8)
                            return false;
                       }
                    }
```

You can of course imagine a helper for this pattern.

The parser then spits out that max/min version along with the tree, the IDE can hold onto that and use that later. This would mean we can reuse trees as long as the user didn't actually use a feature that mattered across them. It would mean the various places we make langversion specific parsing decisions would need an update like this, but if that's reasonably easy to enumerate and a pattern like this isn't too much of a burden going forward, this I think would work.

Now it's a bit strange since the only tree reuse that we're blocking in this case is a scenario where the customer somehow managed to get it to parse in both targets _and_ it has meaning in both targets. Now I'm sure there are plenty of terrible code examples @jaredpar is going to tell me where that's possible, but I'd imagine that's quite rare to mean in practice we'd always share. But for the crazy edge case we still guarantee correctness.

**Comment by jaredpar** [comment]
> The parser then spits out that max/min version along with the tree, the IDE can hold onto that and use that later.

This is a feature we've actually discussed before for different reasons. Essentially it would be good to know what language version was _effectively_ used by the code vs. what maximum language was allowed. That would give us a better sense of how fast items are adopted.

**Comment by jasonmalinowski** [comment]
@jaredpar We could totally aggregate that data and send it up if you wanted the telemetry, so that'd be another good argument for it.

@CyrusNajmabadi For our syntax tree indicies, I know you made changes so they could be reused across debug/release switches -- do we also try to share them across multiple targets? Or do we hash the parse options into that?

**Comment by CyrusNajmabadi** [comment]
We only hash the pp-defines into it. Since those are the only things we think could change the set if symbols indexed in the file.  Technically the langversion could break that. Butt only in pathological files. So we explicitly ignore that and document why.

**Comment by CyrusNajmabadi** [comment]
One thing we don't do is provide a way to look up by checksum. That would be handy so that we could share our indices across projects as well. I'll make a note of that.

**Comment by jasonmalinowski** [comment]
Moving this to @CyrusNajmabadi since he's actually doing it right now.

**Comment by CyrusNajmabadi** [comment]
Done with https://github.com/dotnet/roslyn/pull/66003.
